src/sample/trainers/forward_dynamics.py

```diff
@@ -0,0 +1,160 @@
+        """
+        super().on_data_users_attached()
+        self.forward_dynamics_data_user: DataUser[Tensor] = self.get_data_user(
+            self.data_user_name
+        )
+
+    @override
+    def on_training_models_attached(self) -> None:
+        """Set up model references when they are attached to the trainer.
+
+        This method is called automatically by the PAMIQ framework when
+        training models are attached to the trainer. It retrieves and
+        stores references to the ForwardDynamics model for convenient
+        access during training.
+        """
+
+        super().on_training_models_attached()
+        self.forward_dynamics = self.get_torch_training_model(
+            ModelName.FORWARD_DYNAMICS, ForwardDynamics
+        )
+
+    @override
+    def create_optimizers(self) -> OptimizersSetup:
+        """Create optimizers for the ForwardDynamics model. This method is
+        called automatically by the PAMIQ framework to set up optimizers for
+        the training process. It uses the `partial_optimizer` function to
+        create an optimizer for the ForwardDynamics model's parameters.
+
+        Returns:
+            Dictionary mapping optimizer name to configured optimizer instance.
+        """
+        return {
+            OPTIMIZER_NAME: self.partial_optimizer(
+                self.forward_dynamics.model.parameters()
+            )
+        }
+
+    @override
+    def train(self) -> None:
+        """Execute ForwardDynamics training process.
+
+        This method implements the core ForwardDynamics training loop:
+        1. Creates a dataset and dataloader
+        2. For each batch:
+            - Moves data to the appropriate device
+            - Splits observations, actions, and hidden states
+            - Computes the next observation distribution
+            - Calculates the loss
+            - Backpropagates the loss
+            - Updates the model parameters
+        3. Logs the loss to MLflow
+        4. Increments the global step counter
+        """
+
+        data = self.forward_dynamics_data_user.get_data()
+        dataset = TensorDataset(
+            torch.stack(list(data[DataKey.OBSERVATION])),
+            torch.stack(list(data[DataKey.ACTION])),
+            torch.stack(list(data[DataKey.HIDDEN])),
+        )
+        sampler = self.partial_sampler(dataset=dataset)
+        dataloader = self.partial_dataloader(dataset=dataset, sampler=sampler)
+        device = get_device(self.forward_dynamics.model)
+
+        for _ in range(self.max_epochs):
+            batch: tuple[Tensor, Tensor, Tensor]
+            for batch in dataloader:
+                observations, actions, hiddens = batch
+                observations = observations.to(device)
+                actions = actions.to(device)
+                hiddens = hiddens.to(device)
+
+                observations, hidden, actions, observation_next = (
+                    observations[:, :-1],
+                    hiddens[:, 0],
```

Comment:
`hiddens`は割とテンソルサイズがでかいので 使う分だけ deviceに転送すると良いかと。まあ正直ほとんど変わらないと思うので、今の書き方の方がわかりやすく、直さなくて良いと思いますが
```suggestion

                observations, hidden, actions, observation_next = (
                    observations[:, :-1],
                    hiddens[:, 0].to(device),
```